crates/astria-celestia-client/src/client.rs

```diff
@@ -0,0 +1,306 @@
+use async_trait::async_trait;
+use celestia_rpc::BlobClient;
+use celestia_types::{
+    blob::SubmitOptions,
+    nmt::Namespace,
+    Blob,
+    Commitment,
+};
+use ed25519_consensus::SigningKey;
+use sequencer_types::{
+    RawSequencerBlockData,
+    SequencerBlockData,
+};
+use sequencer_validation::IndexOutOfBounds;
+
+use crate::{
+    blob_space::{
+        celestia_namespace_v0_from_hashed_bytes,
+        SequencerNamespaceData,
+        SignedNamespaceData,
+    },
+    RollupNamespaceData,
+};
+
+impl CelestiaClientExt for jsonrpsee::http_client::HttpClient {}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SubmitSequencerBlocksError {
+    #[error("failed assembling blob for block at index `{index}`")]
+    AssembleBlobs {
+        source: BlobAssemblyError,
+        index: usize,
+    },
+    #[error("the JSONRPC call failed")]
+    JsonRpc(#[source] jsonrpsee::core::Error),
+}
+
+pub struct BadBlob {
+    pub reason: BadBlobReason,
+    pub commitment: Commitment,
+}
+
+pub enum BadBlobReason {
+    Deserialization(serde_json::Error),
+    WrongNamespace(Namespace),
+}
+
+pub struct GetSequencerDataResponse {
+    pub height: u64,
+    pub namespace: Namespace,
+    pub datas: Vec<SignedNamespaceData<SequencerNamespaceData>>,
+    pub bad_blobs: Vec<BadBlob>,
+}
+
+pub struct GetRollupDataResponse {
+    pub height: u64,
+    pub namespace: Namespace,
+    pub datas: Vec<SignedNamespaceData<RollupNamespaceData>>,
+    pub bad_blobs: Vec<BadBlob>,
+}
+
+#[async_trait]
+pub trait CelestiaClientExt: BlobClient {
+    /// Fetch sequencer data at the given height with the provided namespace.
+    ///
+    /// Returns successfully deserialized blobs in the `.datas` field. The
+    /// `.bad_blobs` field contains the celestia commitment for each blob
+    /// that could not be turned into sequencer data and the reason for it.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the underlying `blob.GetAll` JSONRPC failed.
+    async fn get_sequencer_data(
+        &self,
+        height: u64,
+        namespace: Namespace,
+    ) -> Result<GetSequencerDataResponse, jsonrpsee::core::Error> {
+        let blobs = self.blob_get_all(height, &[namespace]).await?;
+
+        let mut datas = Vec::new();
+        let mut bad_blobs = Vec::new();
+        for blob in blobs {
+            if blob.namespace != namespace {
+                bad_blobs.push(BadBlob {
+                    reason: BadBlobReason::WrongNamespace(blob.namespace),
+                    commitment: blob.commitment,
+                });
+            }
+            match serde_json::from_slice(&blob.data) {
+                Ok(data) => datas.push(data),
+                Err(err) => bad_blobs.push(BadBlob {
```

Comment:
we can probably ignore this in most cases, since someone else can just push some random blob to the same namespace, but i guess good to keep in case we have deserialization issues

Reply:
My thinking was that we should start thinking about namespaces becoming occupied by 3rd parties. We could construct metrics around this, and it's just interesting data.